pkg/container/versioned/value.go

```diff
@@ -0,0 +1,337 @@
+		prev = &node.next
+	}
+	if node == nil {
+		// Add the new value at the end
+		prev.Store(newNode)
+	}
+}
+
+// RemoveValueAtVersion changes the validity of the stored value valid at 'version' to have ended at
+// 'version'.
+// 'version' must be later than any the current version visible to the readers.
+// Must be called from a single writer!
+func RemoveValueAtVersion[T any](v *Value[T], version Version) {
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if version < node.versions.first {
+			stacktrace := hclog.Stacktrace()
+			log.Errorf("DeleteValueAtVersion may not be called with version lower than existing already! (%d < %d). Stacktrace: %s", version, node.versions.first, stacktrace)
+		}
+
+		if node.versions.Contains(version) {
+			// truncate the validity of this node to end at 'version'
+			node.versions.past.Store(version)
+			break
+		}
+	}
+}
+
+// Cleaner removes all values whose validity ends at or before 'version'.
+//
+// Must be called from a single writer!
+func Cleaner[T any](v *Value[T], keepVersion Version) {
+	// find all values that are no longer valid at 'version'
+	node := v.head.Load()
+	for node != nil && node.versions.past.Load() <= keepVersion {
+		// This node is no longer visible for readers who are all version 'keepVersion' or
+		// later.
+		node = node.next.Load()
+		v.head.Store(node)
+	}
+}
+
+// GetValue returns value of type 'T' valid for the given handle, or an empty value if none is
+// found.
+func GetValue[T any](v *Value[T], handle Handle) T {
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if node.versions.Contains(handle.getVersion()) {
+			return node.value
+		}
+	}
+	var empty T
+	return empty
+}
+
+// Pair is a struct of a version and any type T
+type Pair[T any] struct {
+	version Version
+	value   T
+}
+
+// PairSlice is a slice of Pairs
+type PairSlice[T any] []Pair[T]
+
+// NewPairSlice returns a new PairSlice of given 'capacity'
+func NewPairSlice[T any](capacity int) PairSlice[T] {
+	return make(PairSlice[T], 0, capacity)
+}
+
+// AppendPair appends a pair of 'version' and 'value' to PairSlice 's', returning 's'
+func AppendPair[T any](s PairSlice[T], version Version, value T) PairSlice[T] {
+	s = append(s, Pair[T]{
+		version: version,
+		value:   value,
+	})
+	return s
+}
+
+// ForEachUpToVersion traverses the Pairs in PairSlice 's', calling function 'f(T)' with each value
+// associated with a values earlier than and including 'version'.
+// Returns the number of calls made.
+func ForEachUpToVersion[T any](s PairSlice[T], version Version, f func(T)) (n int) {
+	for n = 0; n < len(s); n++ {
+		if s[n].version > version {
+			break
+		}
+		f(s[n].value)
+	}
+	return n
+}
+
+// TrimFrontPairSlice trims the first 'n' Pairs off of the PairSlice 's', returning 's'.
+// If 's' gets empty it is reallocated if its capacity exceeds 'capacity'.
+func TrimFrontPairSlice[T any](s PairSlice[T], n int, capacity int) PairSlice[T] {
```

Comment:
TIL, I thought large capacity would stick forever! Do you have a reference for this?

Reply:
Sorry, to be clear - Go cannot release the underlying array while there is a slice referencing it. In that sense, a large cap will stick around, even if `len` is short.

I don't yet understand the usage pattern for `PairSlices` well enough to get why this would be benefitial; though. If we're continuously writing towards at the tail and releasing at the front, should we be using a ring buffer?

Reply:
Do we have a nice "slice like ring buffer", It would need to expand when needed, as hitting a capacity limit is not acceptable.

Maybe we should just always copy the content to the beginning of the slice?

Or maybe just holding on to the storage should be fine? It should get empty very often, and then the runtime should start filling it from the beginning of the storage, correct?